### PR TITLE
security: coordinate DCP Top-K teardown and peer failure

### LINKED ISSUE #181: security: coordinate IPC unmap completion before freeing Top-K exports
## Summary

DCP Top-K cleanup suppresses peer unmap/free failures and may free exported allocations while another rank retains a usable imported pointer.

**Audited revision:** `954fd0174c49502b62547a01f09c404029e6035a`

## Threat model / reachability

A rank errors, exits, or cleans up out of order while peers are active or cleanup calls fail.

## Evidence

- [`b12x/comm/pcie/pcie_dcp_topk.py:176-197`](https://github.com/local-inference-lab/b12x/blob/954fd0174c49502b62547a01f09c404029e6035a/b12x/comm/pcie/pcie_dcp_topk.py#L176-L197) suppresses cleanup errors.
- Without teardown acknowledgements, the exporter cannot know every imported pointer is unused.
- Stale mappings can survive into later work.

## Impact

Use-after-free/stale peer access, context failure, silent corruption, and teardown deadlock/leaks.

## Recommended fix

- Stop launches and synchronize streams.
- Collectively acknowledge unmap completion before freeing exports.
- Propagate cleanup failure into shared state.
- Test forced failures and out-of-order close.

## Verification notes

Source review at the audited revision; runtime exploitation was not attempted unless stated below.

### PR DESCRIPTION
**Pull request by malaiwah**
Closes #181.

Coordinates DCP Top-K preflight, peer-failure propagation, lifecycle transitions, and teardown so asymmetric rank failures cannot strand peer mappings or deadlock later collectives.

Verification: focused lifecycle tests and static checks passed.

## Summary by CodeRabbit

* **New Features**
  * Added support for registering and replaying CUDA graphs with PCIe DCP Top-K operations.
  * Added a retry option for completing resource cleanup after a transient failure.

* **Bug Fixes**
  * Improved validation to detect configuration mismatches before resources are allocated.
  * Strengthened channel lifecycle handling, including coordinated shutdown, retryable cleanup, and safer recovery from failed or unavailable runtimes.
  * Improved protection against resource leaks during interrupted or asymmetric shutdowns.

* **Tests**
  * Added comprehensive coverage for configuration validation, graph execution, cleanup, concurrency, and failure scenarios.

### CODE REVIEW

**Comment by malaiwah** [comment]
Hardware: 2× NVIDIA RTX PRO 6000 Blackwell Server Edition (97,887 MiB each), driver 595.58.03, PCIe SYS topology, peer access enabled. Environment: torch 2.13.0+cu130, CUDA 13.0, CUTLASS DSL 4.6.0. CUDA_VISIBLE_DEVICES=0,1.

GPU-hosted evidence at `2aa6d98fa5b313843df453fa5d4c2560abf8509e`:

```
python -m pytest -q -rs --durations=20 tests/comm/test_pcie_dcp_topk.py tests/comm/test_pcie_dcp_topk_lifecycle.py
```

Result: **29 passed, 17 failed in 81.75s** (wall interval 87.3s). Concrete failures include: the mismatch-test `replacements` object is a tuple rather than a dict, so all nine single-field negative tests raise `KeyError`; two-rank close/retry tests time out and report failure to exchange quiescence status; asymmetric free does not free the expected export; concurrent close fails; and an incompletely initialized object raises an unraisable `_owned_buffers` `AttributeError` in `__del__`. Three graph-capture tests also fail importing `Float32` from CUTLASS despite installed CUTLASS DSL 4.6.0. The lifecycle and negative evidence is not green; this PR is not merge-ready.

**Comment by malaiwah** [comment]
Final two-GPU verification at `3738e02466c3636f874bb68aa6277ca6bd246394`.

Hardware: 2× NVIDIA RTX PRO 6000 Blackwell Server Edition (97,887 MiB each), UUIDs `GPU-468919c3-f392-367d-01a9-2de57c222f25` and `GPU-c2a89142-c177-b92a-1366-d58822694f0a`; driver 595.58.03; PCIe `SYS` topology; peer access enabled. Environment: torch 2.13.0+cu130, CUDA 13.0, CUTLASS DSL 4.6.0. `CUDA_VISIBLE_DEVICES=0,1`.

Commands and results:

```
python -m pytest -q tests/comm/test_pcie_dcp_topk.py tests/comm/test_pcie_dcp_topk_lifecycle.py
# 48 passed in 1.30s

B12X_RUN_PCIE_DCP_TOPK_TEST=1 B12X_PCIE_DCP_TOPK_TP=2 B12X_PCIE_DCP_TOPK_DCP=2 \
python -m pytest -q tests/comm/test_pcie_dcp_topk_gpu.py::test_pcie_dcp_topk_exact_owner_exchange
# 1 passed in 8.64s
```

UTC wall interval: `2026-08-15T04:54:52.947051675Z` to `2026-08-15T04:55:06.836810091Z`.

The published head fixes the parameterized contract tests, multi-rank fake coordinator, partial-construction destructor, CUTLASS test bootstrap, graph registration enforcement, and hidden concurrent-close errors. The GPU test now requires the real CUTLASS DSL and replays through the registered channel gate. Remote IPC ownership without an exchange group is rejected during construction, before a live channel exists; distributed teardown collectives remain bounded by the process group's configured timeout (a rank-local wrapper cannot safely cancel an in-flight NCCL collective). The real two-rank eager, capture, seven queued graph replays, oracle equality, and coordinated close path passes.

File: tests/comm/test_pcie_dcp_topk.py [hunk 1]
```diff
@@ -211,3 +211,213 @@ def test_configuration_rejects_invalid_capacity_and_topk():
...
+        "wire_codec_score": 14,
+        "topology_world": 1,
+        "schedule_max_blocks": 15,
+        "schedule_signal_bytes": 16,
+    }
+    if field not in replacements:
+        raise KeyError(f"unknown field {field}")
+    idx = replacements[field]
+    assert 0 <= idx < len(canonical), f"field {field} index out of range"
+    return idx
```
[the remark this answers is not in the document]
Comment (malaiwah) [reply tests/comm/test_pcie_dcp_topk.py]: Fixed at `3738e02`: the runtime exports `_DCP_TOPK_CONTRACT_FIELDS`, and tests resolve names through that schema while asserting schema/value lengths remain aligned.

File: b12x/comm/pcie/pcie_dcp_topk.py [hunk 2]
```diff
@@ -196,23 +446,175 @@ def _free_ipc_exports(self) -> None:
...
+            if any(free_status):
+                _raise_coordinated_close_error(
+                    "IPC export free",
+                    free_status,
+                    local_errors=free_errors,
+                    exports_retained=False,
+                )
+
+            if self.exchange_group is not None:
+                dist.barrier(group=self.exchange_group)
```
[the remark this answers is not in the document]
Comment (malaiwah) [reply b12x/comm/pcie/pcie_dcp_topk.py:606]: Addressed at `3738e02`: a channel cannot acquire remote IPC ownership without an exchange group; construction rejects it before the channel becomes live. Teardown collectives remain bounded by the process group configured timeout. A rank-local timeout wrapper was intentionally not added because it cannot safely cancel an in-flight NCCL collective; the process-group watchdog is the supported cancellation boundary.

File: b12x/comm/pcie/pcie_dcp_topk.py [hunk 3]
```diff
@@ -408,41 +879,84 @@ def stage_candidates(
...
+                )
+            self._registered_graphs.add(id(graph))
+
+    def replay(self, graph: torch.cuda.CUDAGraph) -> None:
+        """Replay a registered CUDA graph under the lifecycle gate.
+
+        Close cannot quiesce or free while a replay is in flight.
+        """
+        with self._replay_gate():
+            graph.replay()
```
[the remark this answers is not in the document]
Comment (malaiwah) [reply b12x/comm/pcie/pcie_dcp_topk.py:995]: Fixed at `3738e02`: `replay()` rejects unregistered graphs, unit tests cover rejection and registered replay, and the real two-GPU test registers then replays through the channel gate.

File: tests/comm/conftest.py [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+    cutlass.cutlass_dsl.dsl_user_op = lambda f: f
+    sys.modules["cutlass"] = cutlass
+    sys.modules["cutlass.cute"] = cutlass.cute
+    sys.modules["cutlass._mlir"] = cutlass._mlir
+    sys.modules["cutlass._mlir.dialects"] = cutlass._mlir.dialects
+    sys.modules["cutlass._mlir.dialects.llvm"] = cutlass._mlir.dialects.llvm
+    sys.modules["cutlass.cutlass_dsl"] = cutlass.cutlass_dsl
+
+
+_stub_cutlass()
```
[the remark this answers is not in the document]
Comment (malaiwah) [reply tests/comm/conftest.py:73]: Fixed at `3738e02`: the bootstrap probes the complete required DSL module surface, marks fallback installation, exposes `require_real_cutlass`, and the GPU test requires that fixture. The final two-GPU test ran against CUTLASS DSL 4.6.0.

File: tests/comm/test_pcie_dcp_topk_lifecycle.py [hunk 5]
```diff
@@ -0,0 +1,758 @@
...
+    t0 = threading.Thread(target=close_twice)
+    t1 = threading.Thread(target=close1)
+    t0.start()
+    t1.start()
+    t0.join(timeout=10)
+    t1.join(timeout=10)
+
+    assert all(e is None for e in errs)
+    assert c0._lifecycle_state == _ChannelLifecycle.CLOSED
+    assert c1._lifecycle_state == _ChannelLifecycle.CLOSED
```
[the remark this answers is not in the document]
Comment (malaiwah) [reply tests/comm/test_pcie_dcp_topk_lifecycle.py:726]: Fixed at `3738e02`: the rank-1 exception is captured and asserted with diagnostics; the unused `suppress` import was removed.

**Review by lukealonso — CHANGES_REQUESTED** [review]
The lifecycle gate is not enforced when callers retain and directly replay a captured graph, so teardown safety can still be bypassed. At the same time, every staged path takes host locks, adding overhead without fully closing the unsafe replay route.

Please move replay ownership into the integration boundary that can enforce the lifecycle state, avoid per-stage host locking in steady state, and consolidate this with #204 rather than layering two overlapping protocols. Test real multi-rank capture/replay, direct-replay prevention, concurrent teardown, peer failure, and eager/graph latency.

**Comment by lukealonso** [comment]
Closing rather than merging. This adds more than 500 production lines of DCP Top-K preflight, graph, failure-propagation, quarantine, and retry teardown behavior for asymmetric or compromised-rank failures. That is a broad lifecycle redesign rather than a surgical security fix, with meaningful complexity and no production qualification.
